packages/core-p2p/src/socket-server/controllers/peer.ts

```diff
@@ -0,0 +1,202 @@
+import { DatabaseService } from "@arkecosystem/core-database";
+import { Container, Contracts, Providers, Utils } from "@arkecosystem/core-kernel";
+import { Blocks, Crypto, Interfaces, Managers } from "@arkecosystem/crypto";
+import Hapi from "@hapi/hapi";
+
+import { MissingCommonBlockError } from "../../errors";
+import { isWhitelisted } from "../../utils";
+import { TooManyTransactionsError, UnchainedBlockError } from "../errors";
+import { getPeerConfig } from "../utils/get-peer-config";
+import { mapAddr } from "../utils/map-addr";
+import { Controller } from "./controller";
+import { peerSchemas } from "../schemas/peer";
+
+export class PeerController extends Controller {
+    @Container.inject(Container.Identifiers.PeerStorage)
+    private readonly peerStorage!: Contracts.P2P.PeerStorage;
+
+    @Container.inject(Container.Identifiers.PeerProcessor)
+    private readonly peerProcessor!: Contracts.P2P.PeerProcessor;
+
+    @Container.inject(Container.Identifiers.DatabaseService)
+    private readonly database!: DatabaseService;
+
+    public getPeers(request: Hapi.Request, h: Hapi.ResponseToolkit): Contracts.P2P.PeerBroadcast[] {
+        this.validatePayload(request.payload, peerSchemas.getPeers);
+
+        // Add the peer if not already (on every peer endpoint)
+        this.peerProcessor.validateAndAcceptPeer({ ip: request.info.remoteAddress } as Contracts.P2P.Peer);
+
+        return this.peerStorage
+            .getPeers()
+            .map((peer) => peer.toBroadcast())
+            .sort((a, b) => {
+                Utils.assert.defined<number>(a.latency);
+                Utils.assert.defined<number>(b.latency);
+
+                return a.latency - b.latency;
+            });
+    }
+
+    public async getCommonBlocks(
+        request: Hapi.Request,
+        h: Hapi.ResponseToolkit,
+    ): Promise<{
+        common: Interfaces.IBlockData;
+        lastBlockHeight: number;
+    }> {
+        this.validatePayload(request.payload, peerSchemas.getCommonBlocks);
+
+        // Add the peer if not already (on every peer endpoint)
+        this.peerProcessor.validateAndAcceptPeer({ ip: request.info.remoteAddress } as Contracts.P2P.Peer);
+
+        const commonBlocks: Interfaces.IBlockData[] = await this.database.getCommonBlocks((request.payload as any).ids);
+
+        if (!commonBlocks.length) {
+            throw new MissingCommonBlockError();
+        }
+
+        const blockchain = this.app.get<Contracts.Blockchain.Blockchain>(Container.Identifiers.BlockchainService);
+        return {
+            common: commonBlocks[0],
+            lastBlockHeight: blockchain.getLastBlock().data.height,
+        };
+    }
+
+    public async getStatus(request: Hapi.Request, h: Hapi.ResponseToolkit): Promise<Contracts.P2P.PeerPingResponse> {
+        this.validatePayload(request.payload, peerSchemas.getStatus);
```

Comment:
Could this be moved into a plugin that is responsible for validation? Seems to be duplicated in every controller method.

Reply:
done.